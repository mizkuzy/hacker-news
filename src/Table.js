import React from 'react';
import Button from "./Button";

const Table = ({list, pattern, onDismiss, isSearched}) =>
  <div className="table">
    {
      list.filter(isSearched(pattern))
        .map(item => (
          <div key={item.objectID} className="table-row">
            <span className="col-lg">
              <a href={item.url}>{item.title}</a>
            </span>
            <span className="col-md">
              {item.author}
            </span>
            <span className="col-sm">
              {item.num_comments}
            </span>
            <span className="col-sm">
              {item.points}
            </span>
            <span className="col-sm">
              <Button
                onClick={() => onDismiss(item.objectID)}
                className="button-inline"
              >
              Dismiss
              </Button>
            </span>
          </div>
        ))
    }
  </div>;

export default Table