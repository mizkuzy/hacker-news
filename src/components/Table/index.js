import React, { Component } from "react";
import Button from "../Button";
import PropTypes from "prop-types";
import Sort from "../Sort";
import { sortBy } from "lodash";
import {updateSortKeyAndSortReverse} from "../App/hofUpdateStatefunctions";

export const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, "title"),
  AUTHOR: list => sortBy(list, "author"),
  COMMENTS: list => sortBy(list, "num_comments").reverse(),
  POINTS: list => sortBy(list, "points").reverse()
};

class Table extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortKey: "NONE",
      isSortReverse: false
    };

    this.onSort = this.onSort.bind(this);
  }

  onSort(sortKey) {
    this.setState(updateSortKeyAndSortReverse(sortKey));
  }

  render() {
    const { list, onDismiss } = this.props;
    const { sortKey, isSortReverse } = this.state;

    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

    return (
      <div className="table">
        <div className="table-header">
          <span className="col-lg">
            <Sort
              sortKey={"TITLE"}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
              onSort={this.onSort}
            >
              Title
            </Sort>
          </span>
          <span className="col-md">
            <Sort
              sortKey={"AUTHOR"}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
              onSort={this.onSort}
            >
              Author
            </Sort>
          </span>
          <span className="col-xs">
            <Sort
              sortKey={"COMMENTS"}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
              onSort={this.onSort}
            >
              Comments
            </Sort>
          </span>
          <span className="col-xs">
            <Sort
              sortKey={"POINTS"}
              activeSortKey={sortKey}
              isSortReverse={isSortReverse}
              onSort={this.onSort}
            >
              Points
            </Sort>
          </span>
          <span className="col-xs">Archive</span>
        </div>
        {reverseSortedList.map(item => (
          <div key={item.objectID} className="table-row">
            <span className="col-lg">
              <a href={item.url}>{item.title}</a>
            </span>
            <span className="col-md">{item.author}</span>
            <span className="col-xs">{item.num_comments}</span>
            <span className="col-xs">{item.points}</span>
            <span className="col-sm">
              <Button
                onClick={() => onDismiss(item.objectID)}
                className="button-inline"
              >
                Dismiss
              </Button>
            </span>
          </div>
        ))}
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired
};

export default Table;
