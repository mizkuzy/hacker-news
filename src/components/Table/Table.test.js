import React from 'react';
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import Table from "./index";

describe('Table', () => {
  const props = {
    list: [
      {title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y'},
      {title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z'},
    ],
    onDismiss: () => {
    }
  };

  it('renders', () => {
    const div = document.createElement('div');

    ReactDOM.render(<Table {...props}/>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})