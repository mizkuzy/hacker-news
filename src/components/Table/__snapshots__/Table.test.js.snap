// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      className="col-lg"
    >
      <a>
        1
      </a>
    </span>
    <span
      className="col-md"
    >
      1
    </span>
    <span
      className="col-sm"
    >
      1
    </span>
    <span
      className="col-sm"
    >
      2
    </span>
    <span
      className="col-sm"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="col-lg"
    >
      <a>
        2
      </a>
    </span>
    <span
      className="col-md"
    >
      2
    </span>
    <span
      className="col-sm"
    >
      1
    </span>
    <span
      className="col-sm"
    >
      2
    </span>
    <span
      className="col-sm"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
